File: js/src/maplestory/api/kms/dto/character/characterAbility.ts
import * as base from '../../../common/dto/character/characterAbility';
import {
  CharacterAbilityBody,
  CharacterAbilityInfoBody,
  CharacterAbilityPresetBody,
} from '../../response/character/characterAbilityBody';

/**
 * 캐릭터 어빌리티 정보
 */
export class CharacterAbilityDto extends base.CharacterAbilityDto {
  /**
   * 조회 기준일 (KST, 일 단위 데이터로 시, 분은 일괄 0으로 표기)
   */
  public override date: Date | null;

  /**
   * 어빌리티 등급
   */
  public override abilityGrade: string | null;

  /**
   *  어빌리티 정보
   */
  public override abilityInfo: CharacterAbilityInfoDto[];

  /**
   * 보유 명성치
   */
  public override remainFame: number | null;

  /**
   * 적용 중인 어빌리티 프리셋 번호
   */
  public override presetNo: number | null;

  /**
   * 어빌리티 1번 프리셋 전체 정보
   */
  public override abilityPreset1: CharacterAbilityPresetDto | null;

  /**
   * 어빌리티 2번 프리셋 전체 정보
   */
  public override abilityPreset2: CharacterAbilityPresetDto | null;

  /**
   * 어빌리티 3번 프리셋 전체 정보
   */
  public override abilityPreset3: CharacterAbilityPresetDto | null;

  constructor(obj: CharacterAbilityBody) {
    super();

    const {
      date,
      ability_grade,
      ability_info,
      remain_fame,
      preset_no,
      ability_preset_1,
      ability_preset_2,
      ability_preset_3,
    } = obj;

    this.date = date ? new Date(date) : null;
    this.abilityGrade = ability_grade;
    this.abilityInfo = ability_info.map(
      (info) => new CharacterAbilityInfoDto(info),
    );
    this.remainFame = remain_fame;
    this.presetNo = preset_no;
    this.abilityPreset1 = ability_preset_1
      ? new CharacterAbilityPresetDto(ability_preset_1)
      : null;
    this.abilityPreset2 = ability_preset_2
      ? new CharacterAbilityPresetDto(ability_preset_2)
      : null;
    this.abilityPreset3 = ability_preset_3
      ? new CharacterAbilityPresetDto(ability_preset_3)
      : null;
  }
}

/**
 * 캐릭터 어빌리티 상세 정보
 */
export class CharacterAbilityInfoDto extends base.CharacterAbilityInfoDto {
  /**
   * 어빌리티 번호
   */
  public override abilityNo: string;

  /**
   * 어빌리티 등급
   */
  public override abilityGrade: string;

  /**
   * 어빌리티 옵션 및 수치
   */
  public override abilityValue: string;

  constructor(obj: CharacterAbilityInfoBody) {
    super();

    const { ability_no, ability_grade, ability_value } = obj;

    this.abilityNo = ability_no;
    this.abilityGrade = ability_grade;
    this.abilityValue = ability_value;
  }
}

/**
 * 캐릭터 어빌리티 프리셋 정보
 */
export class CharacterAbilityPresetDto extends base.CharacterAbilityPresetDto {
  /**
   * 프리셋의 어빌리티 등급
   */
  public override abilityPresetGrade: string;

  /**
   * 프리셋의 어빌리티 정보
   */
  public override abilityInfo: CharacterAbilityInfoDto[];

  constructor(obj: CharacterAbilityPresetBody) {
    super();

    const { ability_preset_grade, ability_info } = obj;

    this.abilityPresetGrade = ability_preset_grade;
    this.abilityInfo = ability_info.map(
      (info) => new CharacterAbilityInfoDto(info),
    );
  }
}

File: csharp/MapleStory.OpenAPI/Src/KMS/Dto/Character/CharacterAbility.cs
using Newtonsoft.Json;
using Base = MapleStory.OpenAPI.Common.DTO;

namespace MapleStory.OpenAPI.KMS.DTO
{
    /// <summary>
    /// 캐릭터 어빌리티 정보
    /// </summary>
    public class CharacterAbilityDTO : Base.CharacterAbilityDTO<CharacterAbilityInfoDTO, CharacterAbilityPresetDTO>
    {
        /// <summary>
        /// 조회 기준일 (KST, 일 단위 데이터로 시, 분은 일괄 0으로 표기)
        /// </summary>
        [JsonProperty("date")]
        public override DateTimeOffset? Date
        {
            get
            {
                return _date?.ToOffset(TimeSpan.FromHours(9));
            }
            set
            {
                _date = value;
            }
        }

        private DateTimeOffset? _date;

        /// <summary>
        /// 어빌리티 등급
        /// </summary>
        [JsonProperty("ability_grade")]
        public override string? AbilityGrade { get; set; }

        /// <summary>
        /// 어빌리티 정보 리스트
        /// </summary>
        [JsonProperty("ability_info")]
        public override List<CharacterAbilityInfoDTO> AbilityInfo { get; set; }

        /// <summary>
        /// 보유 명성치
        /// </summary>
        [JsonProperty("remain_fame")]
        public override long? RemainFame { get; set; }

        /// <summary>
        /// 적용 중인 어빌리티 프리셋 번호
        /// </summary>
        [JsonProperty("preset_no")]
        public override int? PresetNo { get; set; }

        /// <summary>
        /// 어빌리티 1번 프리셋 전체 정보
        /// </summary>
        [JsonProperty("ability_preset_1")]
        public override CharacterAbilityPresetDTO? AbilityPreset1 { get; set; }

        /// <summary>
        /// 어빌리티 2번 프리셋 전체 정보
        /// </summary>
        [JsonProperty("ability_preset_2")]
        public override CharacterAbilityPresetDTO? AbilityPreset2 { get; set; }

        /// <summary>
        /// 어빌리티 3번 프리셋 전체 정보
        /// </summary>
        [JsonProperty("ability_preset_3")]
        public override CharacterAbilityPresetDTO? AbilityPreset3 { get; set; }
    }

    /// <summary>
    /// 캐릭터 어빌리티 상세 정보
    /// </summary>
    public class CharacterAbilityInfoDTO : Base.CharacterAbilityInfoDTO
    {
        /// <summary>
        /// 어빌리티 번호
        /// </summary>
        [JsonProperty("ability_no")]
        public override string AbilityNo { get; set; }

        /// <summary>
        /// 어빌리티 등급
        /// </summary>
        [JsonProperty("ability_grade")]
        public override string AbilityGrade { get; set; }

        /// <summary>
        /// 어빌리티 옵션 및 수치
        /// </summary>
        [JsonProperty("ability_value")]
        public override string AbilityValue { get; set; }
    }

    /// <summary>
    /// 캐릭터 어빌리티 프리셋 정보
    /// </summary>
    public class CharacterAbilityPresetDTO : Base.CharacterAbilityPresetDTO<CharacterAbilityInfoDTO>
    {
        /// <summary>
        /// 프리셋의 어빌리티 등급
        /// </summary>
        [JsonProperty("ability_preset_grade")]
        public override string AbilityPresetGrade { get; set; }

        /// <summary>
        /// 프리셋의 어빌리티 정보
        /// </summary>
        [JsonProperty("ability_info")]
        public override List<CharacterAbilityInfoDTO> AbilityInfo { get; set; }
    }
}
